projects/lights_control/src/lights_gpio.c

```diff
@@ -2,211 +2,113 @@
 #include "event_queue.h"
 #include "gpio.h"
 #include "gpio_mcu.h"
+#include "lights_config.h"
 #include "lights_events.h"
 #include "status.h"
 
-#define LIGHTS_HORN_PERIPHERAL 0
-#define NUM_LIGHTS_HORN_PERIPHERALS 1
-#define LIGHTS_STROBE_PERIPHERAL 0
-#define NUM_LIGHTS_STROBE_PERIPHERALS 1
-
-typedef enum {
-  LIGHTS_HIGH_BEAMS_RIGHT_PERIPHERAL = 0,
-  LIGHTS_HIGH_BEAMS_LEFT_PERIPHERAL,
-  NUM_LIGHTS_HIGH_BEAMS_PERIPHERALS
-} LightsHighBeamsPeripherals;
-
-typedef enum {
-  LIGHTS_LOW_BEAMS_RIGHT_PERIPHERAL = 0,
-  LIGHTS_LOW_BEAMS_LEFT_PERIPHERAL,
-  NUM_LIGHTS_LOW_BEAMS_PERIPHERALS
-} LightsLowBeamsPeripherals;
-
-typedef enum {
-  LIGHTS_DRL_RIGHT_PERIPHERAL = 0,
-  LIGHTS_DRL_LEFT_PERIPHERAL,
-  NUM_LIGHTS_DRL_PERIPHERALS
-} LightsDRLPeripherals;
-
-typedef enum {
-  LIGHTS_RIGHT_REAR_BRAKE_PERIPHERAL = 0,
-  LIGHTS_RIGHT_REAR_OUTER_BRAKE_PERIPHERAL,
-  LIGHTS_LEFT_REAR_BRAKE_PERIPHERAL,
-  LIGHTS_LEFT_REAR_OUTER_BRAKE_PERIPHERAL,
-  LIGHTS_CENTRE_BRAKE_PERIPHERAL,
-  NUM_LIGHTS_BRAKES_PERIPHERALS
-} LightsBrakesPeripherals;
-
-typedef enum {
-  LIGHTS_SIDE_LEFT_INDICATOR_PERIPHERAL = 0,
-  LIGHTS_LEFT_TURN_PERIPHERAL,
-  NUM_LIGHTS_FRONT_SIGNAL_LEFT_PERIPHERALS
-} LightsFrontSignalLeftPeripherals;
-
-typedef enum {
-  LIGHTS_SIDE_RIGHT_INDICATOR_PERIPHERAL = 0,
-  LIGHTS_RIGHT_TURN_PERIPHERAL,
-  NUM_LIGHTS_FRONT_SIGNAL_RIGHT_PERIPHERALS
-} LightsFrontSignalRightPeripherals;
-
-typedef enum {
-  LIGHTS_LEFT_REAR_OUTER_TURN_PERIPHERAL = 0,
-  LIGHTS_LEFT_REAR_TURN_PERIPHERAL,
-  NUM_LIGHTS_REAR_SIGNAL_LEFT_PERIPHERALS
-} LightsRearSignalLeftPeripherals;
-
-typedef enum {
-  LIGHTS_RIGHT_REAR_OUTER_TURN_PERIPHERAL = 0,
-  LIGHTS_RIGHT_REAR_TURN_PERIPHERAL,
-  NUM_LIGHTS_REAR_SIGNAL_RIGHT_PERIPHERALS
-} LightsRearSignalRightPeripherals;
-
-static const GPIOAddress s_horn_peripheral_addresses[NUM_LIGHTS_HORN_PERIPHERALS] = {
-  [LIGHTS_HORN_PERIPHERAL] = { .port = GPIO_PORT_B, .pin = 11 },  //
-};
-
-static const GPIOAddress s_high_beams_peripheral_addresses[NUM_LIGHTS_HIGH_BEAMS_PERIPHERALS] = {
-  [LIGHTS_HIGH_BEAMS_RIGHT_PERIPHERAL] = { .port = GPIO_PORT_B, .pin = 1 },  //
-  [LIGHTS_HIGH_BEAMS_LEFT_PERIPHERAL] = { .port = GPIO_PORT_B, .pin = 15 },  //
-};
-
-static const GPIOAddress s_low_beams_peripheral_addresses[NUM_LIGHTS_LOW_BEAMS_PERIPHERALS] = {
-  [LIGHTS_LOW_BEAMS_RIGHT_PERIPHERAL] = { .port = GPIO_PORT_B, .pin = 2 },  //
-  [LIGHTS_LOW_BEAMS_LEFT_PERIPHERAL] = { .port = GPIO_PORT_A, .pin = 8 },   //
+static const GPIOAddress *s_addresses;
+
+// set to an invalid value, gets set when module gets initialized
+static LightsBoard s_board = NUM_LIGHTS_BOARDS;
+
+// mapping events to their corresponding peripherals using bitsets , index of bit corresponds to
+// the index of GPIO addresses in lights_config.c
+static uint16_t s_front_event_mappings[NUM_FRONT_LIGHTS_EVENTS] = {
+  [FRONT_LIGHTS_EVENT_HORN] = 0x001,           //
```

Comment:
Please use your enums instead of magic numbers. I suggest creating a header to contain your bitsets if you want to avoid them becoming massive.

Reply:
like do something like this? If I have something like: 
- `FRONT_LIGHTS_3`
- `FRONT_LIGHTS_4`

that both correspond to `event_1`, then to be able to make a bitset I'm gonna have to do something like this: 
```
static uint16_t s_front_event_mappings[NUM_FRONT_LIGHTS_EVENTS] = {
//
  [event_1] = (FRONT_LIGHTS_3 << 1) | (FRONT_LIGHTS_4 << 1),
//
}
```

because enums go up by 1, but in my bitsets every bit corresponds to a light.

Reply:
Yeah, but that's more explicit than a magic number that we'd have to extract. You could always use macros to make it neater.

Reply:
yay macros ok

Reply:
I would probably just set the values in the enum itself so `FRONT_LIGHTS_3 = 1 << x`. Then you could use `FRONT_LIGHTS_3 | FRONT_LIGHTS_4`.

Reply:
one problem that would have is that when declaring the addresses in `lights_config.c`, it'll  initialize the addresses array with lots of 0 entries.

Reply:
You could just use a different set of enums or use standard array notation.